MPI_Builds/sample_sort.cpp

```diff
@@ -0,0 +1,355 @@
+                    final_array[total_count] = received_array[j];
+                    total_count++;
+                }
+            }
+        }
+
+        // send mini finalized arrays back to other processes
+        for (int i=1; i<p; i++) {
+            CALI_MARK_BEGIN("comm");
+            CALI_MARK_BEGIN("comm_large"); // send all to other processes
+            MPI_Send(&final_array[i*local_size], local_size, MPI_INT, i, FROM_MASTER, MPI_COMM_WORLD);   
+            CALI_MARK_END("comm_large");
+            CALI_MARK_END("comm");
+        }
+        // for process 0
+        for (int i=0; i<local_size; i++) {
+            final_subarray[i] = final_array[i];
+        }
+
+        // print final output
+        // for (int i=0; i<n; i++) {
+        //     printf("final[%d] : %d\n", i, final_array[i]);
+        // }
+
+        // process 0 returns sorted bucket 0, bucket 1, ..., bucket `p-1`
+    }
+
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    if (taskid != 0) {
+        CALI_MARK_BEGIN("comm");
+        CALI_MARK_BEGIN("comm_large"); // size n/p
+        MPI_Recv(&final_subarray, local_size, MPI_INT, 0, FROM_MASTER, MPI_COMM_WORLD, &status);
+        CALI_MARK_END("comm_large");
+        CALI_MARK_END("comm");
+
+    }
+
+    // for (int i=0; i<local_size; i++) {
+    //     printf("final[%d] : %d, process : %d\n", i, final_subarray[i], taskid);
+    // }
+
+    CALI_MARK_BEGIN("correctness_check");
+    rc = verify_sort(final_subarray, local_size, 0);
+    CALI_MARK_END("correctness_check");
+
+    if (rc) {
+        printf("Process %d: Sort check failed.\n", taskid);
+    } else {
+        printf("Process %d: Sort check succeeded.\n", taskid);
+    }
+
+adiak::init(NULL);
+    adiak::launchdate();    // launch date of the job
+    adiak::libraries();     // Libraries used
+    adiak::cmdline();       // Command line used to launch the job
+    adiak::clustername();   // Name of the cluster
+    adiak::value("algorithm", "sample"); // The name of the algorithm you are using (e.g., "merge", "bitonic")
+    adiak::value("programming_model", "mpi"); // e.g. "mpi"
+    adiak::value("data_type", "int"); // The datatype of input elements (e.g., double, int, float)
+    adiak::value("size_of_data_type", sizeof(int)); // sizeof(datatype) of input elements in bytes (e.g., 1, 2, 4)
+    adiak::value("input_size", n); // The number of elements in input dataset (1000)
+    adiak::value("input_type", argv[2]); // For sorting, this would be choices: ("Sorted", "ReverseSorted", "Random", "1_perc_perturbed")
+    adiak::value("num_procs", p); // The number of processors (MPI ranks)
+    adiak::value("scalability", "strong"); // The scalability of your algorithm. choices: ("strong", "weak")
+    adiak::value("group_num", 23); // The number of your group (integer, e.g., 1, 10)
+    adiak::value("implementation_source", "handwritten"); // Where you got the source code of your algorithm. choices: ("online", "ai", "handwritten")

```

Comment:
"Handwritten" doesn't quite fit with the wikipedia link at the top of `quick_sort.h`, though I think it's technically fine because it's functionally a black box called by your sample sort and sequential sorting is not the assignment's focus.